### DISCUSSION TITLE
0025 XLS-25d: Enhanced Secret Numbers (XLS-12 Extension)

### DISCUSSION THREAD

**Discussion by nbougalis** [XRPLF/XRPL-Standards#63] [XLS Proposals]
## The earlier version can be found by looking at the edit history.

**Comment by WietseWind** [comment]
Great proposal @nbougalis, thank you. I agree with you on the less than optimal checksum for the existing Secret numbers, and the combined ed+secp compatibility is a great idea as well.

When I find some spare time I will work on an implementation in https://github.com/WietseWind/xrpl-secret-numbers/ :+1:

**Comment by nbougalis** [comment]
@WietseWind's comment above stems from a conversation in the "Technical" channel on Mattermost. He wrote:

>Nice Nik B!! Very nice. I'll think about how this is going to be used in the UI, allow people to enter and then at the end tap a button "I have more numbers" or something like that
>I'll add support on a branch of https://www.npmjs.com/package/xrpl-secret-numbers when I find a moment

I responded that:

>[a] small tweak can make it trivial... under the existing spec, by construction block 'A'  can never end with a 9 (it's `(x * 10) + (x % 9)`) so we can leverage that to construct the A block for XLS25 in a special way: `A = (x * 10) + 9`. You lose the "direct" checksum on the A block but the overall checksum would still catch the error.

I thought about it some more and there might be a further improvement on my potential improvement. By construction, no valid XLS12 block can have a value greater than `655356` (i.e. `65535 * 10 + (65535 % 9)`). We can take advantage of that, because it means that any block that begins with a `7`, `8` or `9` cannot be valid XLS12. So instead, consider this:

Again, assume that `C` is the checksum described above and `F` is the flag bitfield:

`X = (F ^ C) % 3`
`A = ((X + 7) * 100000) + (C * 256 + F)`

The end result is an `A` block that begins with `7`, `8` or `9`, which means it cannot be a valid XLS12 block, allowing us to detect the type automatically. Plus, we still retain a checksum in the `A` block, albeit more "fragile" than the remaining blocks.

If you think this is a reasonable approach, I'll make the necessary tweaks to the spec above.

**Comment by mvadari** [comment]
@nbougalis should this spec be moved into PR and kept in the repo or should it be closed? Note that this does not mean that the spec has been accepted/implemented/finalized.

I will move it to PR by copy-pasting the existing draft if there is no response in the next week.
